proposals/0000-unsaturated-type-families.rst

```diff
@@ -0,0 +1,1051 @@
+is invalid, but ``forall m. @U forall x. @m x`` is valid. The same applies to the
+visible case: ```forall m -> @m Type``` is rejected.
+
+.. _Inference:
+
+Meaning of an annotation-free ``->``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Even though this proposal introduces a way to annotate arrows, in many cases the
+annotations can be inferred. The primary aim of inference is to ease the
+transition as most programs written today can be unambiguously inferred.
+
+The meaning of ``(->)`` depends on the context in which it is written. Below is
+a list of the different contexts with examples. Some of these are a result of a
+*defaulting* mechanism, and some are *hard rules*. The first category can be
+overridden with an appropriate annotation, but the second can not. The
+defaulting mechanism first invents a unification variable for the matchability,
+and only applies the defaulting rule here when the variable remains
+unconstrained after constraint solving.
+
+Data types
+##########
+
+The kind arrows of data types (and data families) are all matchable.
+
+::
+
+  -- inferred:  Type -> @M Type
+  type Maybe :: Type -> Type
+  data Maybe a = ...
+
+here, users are not required to specify ``Type -> @M Type``, as this information
+can be inferred from the data declaration itself. This is a *hard rule*, and
+it is an error to write ::
+
+  type Maybe :: Type -> @U Type
+  data Maybe a = ...
+
+Data constructors
+#################
+
+Data constructors are matchable. This means that using either syntax ::
+
+  data Maybe a = Nothing | Just a
+
+  data Maybe a where
+    Nothing :: Maybe a
+    Just :: a -> Maybe a
+
+``Just :: a -> @M Maybe a``. GHC already eta-expands data constructors
+automatically, so writing ``map Just xs`` will work even though ``map`` expects
+an unmatchable argument, because ``Just`` will be elaborated to ``(\x -> Just x)``
+which is an unmatchable lambda. Promoting ``Just`` thus results in a matchable
+constructor ``'Just :: a -> @M Maybe a``.
+
+This is a *hard rule*.
+
+Type families
+#############
+
+Type family (and type synonym) *arguments* are unmatchable
+
+::
+
+  -- inferred: Type -> @U Type
+  type Id :: Type -> Type
+  type Id a = a
+
+is unambiguous, and no annotation is required. However, the unambiguity here
+arises not solely due to the fact that ``Id`` is a type synonym, but also that it
+binds its argument on the left-hand side. This is a *hard rule*.
+
+The arrows not corresponding to arguments
+bound on the LHS are inferred to be matchable (by *default*) ::
+
+  -- inferred: Type -> @M Type
+  type MyMaybe :: Type -> Type
+  type family MyMaybe where
+    MyMaybe = Maybe
+
+``MyMaybe`` is a nullary type family, and its return kind is thus matchable (see
+the *Arity of type families* section for more details about type family arities).
+
+The following is rejected ::
+
+  -- inferred: Type -> @M Type
+  type MyId :: Type -> Type
+  type family MyId where
+    MyId = Id -- rejected
```

Comment:
Aha, judging by [Alternative #6](https://github.com/kcsongor/ghc-proposals/blob/unsaturated-type-families/proposals/0000-unsaturated-type-families.rst#id35), the issue is that we've given `MyId` an explicit kind signature that needs checking and defaulting before we check the definition. That makes sense, but it would be helpful to highlight.

Although.. now I'm confused as to why we're allowed to inspect the binders on the LHS of the definition to determine matchability in the kind signature, but not the RHS.

Reply:
I'll update the text with a note highlighting this behaviour.

To answer the second point: this is what GHC does today, it looks at the LHS to determine the arity and compute the result kind of the TyCon (where the result kind is the original signature minus the bound arguments). This pass gives us an opportunity to figure out the matchabilities. Importantly, this is a very syntactic check, and so it doesn't wreak havoc with kind checking/inference.